### ISSUE TITLE
Accessing the value of the loss from within the v2 optimizer

### ISSUE THREAD

**Issue by ColinMcAuliffe** [keras-team/keras#14062]
- Have I written custom code (as opposed to using example directory):  yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Ubuntu 16.04.5
- TensorFlow backend (yes / no):  yes
- TensorFlow version:  2.2.0
- Keras version:  2.3.1
- Python version:  3.5.2

Hi all, I'm interested in experimenting with an optimization technique that uses the loss history and the value of the loss on the current batch to adapt the learning rate, described here https://arxiv.org/pdf/1802.05074.pdf

I've implemented custom optimizers by inheriting from optimizer_v2, and defining the [_resource_apply_dense](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L1086) method, but it looks like I can not access the loss value for the batch which is needed to adapt the step size for this particular optimizer. As far as I can tell from looking at [the source,](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training.py#L2648) the loss is never passed to the optimizer.

I've also looked into using a callback, but I don't think this will work since the loss is not available for an on_batch_begin callback, and for an on_batch_end callback the loss is available but by that time the gradients have already been applied. 

Are there plans to make the current batch loss value accessible from the optimizer in future versions? Also, can anyone suggest a workaround here? Thank you, Colin

**Comment by qlzh727** [comment]
I am bit confused about the link you provided. The loss value did get passed to optimizers when it does the backprop, see https://github.com/keras-team/keras/blob/42bf9972492f47c3d3c249de9c20942ba217937d/keras/optimizer_v2/optimizer_v2.py#L499.

Assigning to @chenmoneygithub who works on optimizers.

**Comment by chenmoneygithub** [comment]
If you are using a custom training loop, then the loss value is not available inside optimizer.  Besides, even if you are using minimize(), which takes in the current loss, we will not keep the loss history. This behavior is expected, as tracking loss history is out of the scope of optimizer.

There are two ways to handle the learning rate adaption:
1) in your custom training loop, before calling optimizer, you could manually compute the new learning rate, and set the optimizer's learning rate.
2) via callback. You may override on_batch_end so that the learning rate is adapted, though it would lag behind by one step. That is, overriding on_batch_end would adapt the learning rate for next batch based on current batch's loss. This is an approximation method, but should achieve close performance.

Besides, if you want to create a new optimizer, we encourage you to check out the reworked optimizer: [link](https://github.com/keras-team/keras/blob/master/keras/optimizer_experimental/optimizer.py). 

Please let me know if you have any further questions, thanks!

**Comment by pswpswpsw** [comment]
> * Have I written custom code (as opposed to using example directory):  yes
> * OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Ubuntu 16.04.5
> * TensorFlow backend (yes / no):  yes
> * TensorFlow version:  2.2.0
> * Keras version:  2.3.1
> * Python version:  3.5.2
> 
> Hi all, I'm interested in experimenting with an optimization technique that uses the loss history and the value of the loss on the current batch to adapt the learning rate, described here https://arxiv.org/pdf/1802.05074.pdf
> 
> I've implemented custom optimizers by inheriting from optimizer_v2, and defining the [_resource_apply_dense](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L1086) method, but it looks like I can not access the loss value for the batch which is needed to adapt the step size for this particular optimizer. As far as I can tell from looking at [the source,](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training.py#L2648) the loss is never passed to the optimizer.
> 
> I've also looked into using a callback, but I don't think this will work since the loss is not available for an on_batch_begin callback, and for an on_batch_end callback the loss is available but by that time the gradients have already been applied.
> 
> Are there plans to make the current batch loss value accessible from the optimizer in future versions? Also, can anyone suggest a workaround here? Thank you, Colin

Interestingly I am doing exactly the same thing: implementing L4 in tensorflow 2 optimizer API. I also arrived at the same problem as you described. Do you have a chance to fix this?

**Comment by chenmoneygithub** [comment]
@pswpswpsw Hi, this is not a bug, but an expected behavior. Keras optimizer's role is to apply given `gradients` on `model variables`. If you wish to use the loss during gradients back propagation, please write a custom training loop.

However, I do see recently there are more optimization algorithms doing gradients "looking ahead", so we may adapt the optimizer to be aware of the loss in the future.

**Comment by ColinMcAuliffe** [comment]
@pswpswpsw I ended up writing a custom training loop to implement this, which turned out to be a lot easier than I expected :)
